File: internal/plan/plan_test.go
package plan

import (
	"strings"
	"testing"

	"github.com/overkazaf/re-agent/internal/types"
)

func steps(pairs ...string) []types.PlanStep {
	var out []types.PlanStep
	for index := 0; index+1 < len(pairs); index += 2 {
		out = append(out, types.PlanStep{Text: pairs[index], Status: types.PlanStepStatus(pairs[index+1])})
	}
	return out
}

func TestUpdateReportsOnlyRealChanges(t *testing.T) {
	tracker := &Tracker{}
	first := tracker.Update(steps("triage", "in_progress", "solve", "pending"),
		types.PlanUpdateMeta{Source: "codex"})
	if first == nil {
		t.Fatal("first update should produce a snapshot")
	}
	if again := tracker.Update(steps("triage", "in_progress", "solve", "pending"),
		types.PlanUpdateMeta{Source: "update_plan"}); again != nil {
		t.Fatal("an unchanged list should not redraw")
	}
	// The source is still recorded so the snapshot never reports a stale origin.
	if tracker.Current().Source != "update_plan" {
		t.Fatalf("source not updated: %s", tracker.Current().Source)
	}
	changed := tracker.Update(steps("triage", "completed", "solve", "in_progress"),
		types.PlanUpdateMeta{Source: "codex"})
	if changed == nil {
		t.Fatal("a status change should produce a snapshot")
	}
}

func TestUpdateCarriesTimingsAcrossUpdates(t *testing.T) {
	tracker := &Tracker{}
	tracker.Update(steps("triage", "in_progress"), types.PlanUpdateMeta{Source: "x"})
	started := tracker.Current().Steps[0].StartedAt
	if started == 0 {
		t.Fatal("an in_progress step should be stamped with a start")
	}
	snapshot := tracker.Update(steps("triage", "completed"), types.PlanUpdateMeta{Source: "x"})
	if snapshot.Steps[0].StartedAt != started {
		t.Fatal("start time was not carried across the update")
	}
	if snapshot.Steps[0].CompletedAt == 0 {
		t.Fatal("completion was not stamped")
	}
}

func TestSanitizeDropsJunkAndClamps(t *testing.T) {
	tracker := &Tracker{}
	var many []types.PlanStep
	for i := 0; i < 80; i++ {
		many = append(many, types.PlanStep{Text: "step", Status: types.StepPending})
	}
	many = append(many, types.PlanStep{Text: "   ", Status: types.StepPending})
	snapshot := tracker.Update(many, types.PlanUpdateMeta{Source: "x"})
	if len(snapshot.Steps) != maxSteps+1 {
		t.Fatalf("expected the list to be clamped with a marker, got %d", len(snapshot.Steps))
	}
	if !strings.Contains(snapshot.Steps[len(snapshot.Steps)-1].Text, "more steps not shown") {
		t.Fatal("a truncated plan must not look complete")
	}
}

func TestSanitizeStripsControlCharacters(t *testing.T) {
	tracker := &Tracker{}
	snapshot := tracker.Update(
		[]types.PlanStep{{Text: "\x1b[31mred\x1b[0m\tstep\nwith breaks", Status: "bogus"}},
		types.PlanUpdateMeta{Source: "x"})
	step := snapshot.Steps[0]
	if strings.ContainsAny(step.Text, "\x1b\n\t") {
		t.Fatalf("control characters survived: %q", step.Text)
	}
	if step.Status != types.StepPending {
		t.Fatalf("unknown status should fall back to pending, got %s", step.Status)
	}
}

func TestCounts(t *testing.T) {
	done, total := Counts(&types.PlanSnapshot{Steps: steps("a", "completed", "b", "pending", "c", "completed")})
	if done != 2 || total != 3 {
		t.Fatalf("counts wrong: %d/%d", done, total)
	}
	if done, total := Counts(nil); done != 0 || total != 0 {
		t.Fatalf("nil snapshot should count as empty, got %d/%d", done, total)
	}
}

File: src/core/plan.test.ts
import { describe, expect, test } from "bun:test";
import { PlanTracker, planCounts } from "./plan";
import type { PlanStep } from "../types";

const meta = { source: "update_plan" };

/** Junk arrives from CLIs and models alike, so tests feed it unchecked. */
const loose = (steps: unknown): PlanStep[] => steps as PlanStep[];

const ESC = String.fromCharCode(0x1b);
const BELL = String.fromCharCode(0x07);
const NUL = String.fromCharCode(0x00);

describe("PlanTracker.update", () => {
  test("returns a snapshot for the first update", () => {
    const tracker = new PlanTracker();
    const snapshot = tracker.update([{ text: "triage binary", status: "in_progress" }], meta);
    expect(snapshot).toBeDefined();
    expect(snapshot?.steps).toMatchObject([{ text: "triage binary", status: "in_progress" }]);
    // An in_progress step is stamped as started so the HUD can time it.
    expect(snapshot?.steps[0].startedAt).toBeNumber();
    expect(snapshot?.steps[0].completedAt).toBeUndefined();
    expect(snapshot?.source).toBe("update_plan");
    expect(snapshot?.note).toBeUndefined();
    expect(tracker.current).toBe(snapshot);
  });

  test("debounces an identical repeat update", () => {
    const tracker = new PlanTracker();
    const steps: PlanStep[] = [
      { text: "triage binary", status: "completed" },
      { text: "find flag", status: "pending" },
    ];
    const first = tracker.update(steps, meta);
    expect(first).toBeDefined();
    expect(tracker.update([...steps], meta)).toBeUndefined();
    expect(tracker.current).toBe(first);
  });

  test("emits a snapshot when only a status changed", () => {
    const tracker = new PlanTracker();
    tracker.update([{ text: "triage binary", status: "pending" }], meta);
    const changed = tracker.update([{ text: "triage binary", status: "completed" }], meta);
    expect(changed).toBeDefined();
    expect(changed?.steps[0].status).toBe("completed");
  });

  test("emits a snapshot when only the note changed", () => {
    const tracker = new PlanTracker();
    const steps: PlanStep[] = [{ text: "triage binary", status: "pending" }];
    tracker.update(steps, meta);
    const renoted = tracker.update([...steps], { source: "update_plan", note: "waiting on unpack" });
    expect(renoted).toBeDefined();
    expect(renoted?.note).toBe("waiting on unpack");
  });

  test("drops blank and text-less steps", () => {
    const tracker = new PlanTracker();
    const snapshot = tracker.update(
      loose([
        { text: "keep me", status: "pending" },
        { text: "   ", status: "pending" },
        { text: "", status: "completed" },
        { status: "pending" },
        { text: 42, status: "pending" },
        null,
      ]),
      meta,
    );
    expect(snapshot?.steps).toEqual([{ text: "keep me", status: "pending" }]);
  });

  test("ignores an all-blank list and leaves current untouched", () => {
    const tracker = new PlanTracker();
    const first = tracker.update([{ text: "triage binary", status: "pending" }], meta);
    expect(tracker.update(loose([{ text: "  ", status: "pending" }, { text: "" }]), meta)).toBeUndefined();
    expect(tracker.update([], meta)).toBeUndefined();
    expect(tracker.current).toBe(first);
  });

  test("coerces an unknown status to pending", () => {
    const tracker = new PlanTracker();
    const snapshot = tracker.update(
      loose([
        { text: "bogus status", status: "in-progress" },
        { text: "missing status" },
        { text: "numeric status", status: 3 },
      ]),
      meta,
    );
    expect(snapshot?.steps.map(step => step.status)).toEqual(["pending", "pending", "pending"]);
  });

  test("strips ANSI escapes and control characters from step text", () => {
    const tracker = new PlanTracker();
    const noisy = `${ESC}[31mdump${ESC}[0m${BELL} the${NUL}flag`;
    const snapshot = tracker.update([{ text: noisy, status: "pending" }], meta);
    expect(snapshot?.steps[0].text).toBe("dump the flag");
  });

  test("truncates over-long step text", () => {
    const tracker = new PlanTracker();
    const snapshot = tracker.update([{ text: "A".repeat(500), status: "pending" }], meta);
    const text = snapshot?.steps[0].text ?? "";
    expect(text.length).toBe(200);
    expect(text.endsWith("…")).toBe(true);
    expect(text.startsWith("AAA")).toBe(true);
  });

  test("clamps the list at 64 steps and says how many were dropped", () => {
    const tracker = new PlanTracker();
    const steps: PlanStep[] = Array.from({ length: 100 }, (_, i) => ({ text: `step ${i}`, status: "pending" }));
    const snapshot = tracker.update(steps, meta);
    // 64 real steps plus one marker row, so a truncated plan never reads as complete.
    expect(snapshot?.steps).toHaveLength(65);
    expect(snapshot?.steps[63].text).toBe("step 63");
    expect(snapshot?.steps[64].text).toBe("… 36 more steps not shown");
  });

  test("does not add the dropped marker at exactly the clamp", () => {
    const tracker = new PlanTracker();
    const steps: PlanStep[] = Array.from({ length: 64 }, (_, i) => ({ text: `step ${i}`, status: "pending" }));
    expect(tracker.update(steps, meta)?.steps).toHaveLength(64);
  });

  test("carries step timings across updates instead of restamping them", async () => {
    const tracker = new PlanTracker();
    tracker.update([{ text: "triage", status: "in_progress" }, { text: "solve", status: "pending" }], meta);
    const startedAt = tracker.current?.steps[0].startedAt;
    expect(startedAt).toBeNumber();
    expect(tracker.current?.steps[1].startedAt).toBeUndefined();

    await Bun.sleep(5);
    tracker.update([{ text: "triage", status: "completed" }, { text: "solve", status: "in_progress" }], meta);
    // The original start survives, so a completed step can report real elapsed time.
    expect(tracker.current?.steps[0].startedAt).toBe(startedAt!);
    expect(tracker.current?.steps[0].completedAt).toBeNumber();
    expect(tracker.current?.steps[0].completedAt! - startedAt!).toBeGreaterThanOrEqual(4);
    expect(tracker.current?.steps[1].startedAt).toBeNumber();
    expect(tracker.current?.steps[1].completedAt).toBeUndefined();
  });

  test("matches steps by provider id, so a renamed step keeps its timing", () => {
    const tracker = new PlanTracker();
    tracker.update([{ id: "1", text: "triage", status: "in_progress" }], meta);
    const startedAt = tracker.current?.steps[0].startedAt;
    tracker.update([{ id: "1", text: "triage the ELF", status: "in_progress" }], meta);
    expect(tracker.current?.steps[0].text).toBe("triage the ELF");
    expect(tracker.current?.steps[0].startedAt).toBe(startedAt!);
  });

  test("reset clears the current snapshot", () => {
    const tracker = new PlanTracker();
    tracker.update([{ text: "triage binary", status: "pending" }], meta);
    tracker.reset();
    expect(tracker.current).toBeUndefined();
    // A repeat of the pre-reset plan is no longer a no-op.
    expect(tracker.update([{ text: "triage binary", status: "pending" }], meta)).toBeDefined();
  });
});

describe("planCounts", () => {
  test("returns zeroes for an undefined snapshot", () => {
    expect(planCounts(undefined)).toEqual({ done: 0, total: 0 });
  });

  test("counts only completed steps", () => {
    const tracker = new PlanTracker();
    const snapshot = tracker.update(
      [
        { text: "triage", status: "completed" },
        { text: "unpack", status: "completed" },
        { text: "solve", status: "in_progress" },
        { text: "write up", status: "pending" },
      ],
      meta,
    );
    expect(planCounts(snapshot)).toEqual({ done: 2, total: 4 });
  });
});
